### PR TITLE
[AIRFLOW-1836] airflow uses OAuth Provider keycloak

### PR DESCRIPTION
**Pull request by fisher-monkey**
Dear Airflow maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [ ] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "[AIRFLOW-XXX] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-1836

### Description
- [ ] Here are some details about my PR, including screenshots of any UI changes:

### Tests
- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits
- [ ] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

**Comment by Fokko** [comment]
Hi @fisher-monkey, thanks for contributing. Could you open a Jira ticket and add a description about the implementation that you've did?

File: airflow/contrib/auth/backends/keycloak_auth.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+    def login(self, request):
+        _log.debug('Redirecting user to GHE login')
+        return self.ghe_oauth.authorize(callback=url_for(
+            'ghe_oauth_callback',
+            _external=True,
+            next=request.args.get('next') or request.referrer or None))
+
+    def get_ghe_user_profile_info(self, ghe_token):
+        #resp = self.ghe_oauth.get(self.ghe_api_route('/userinfo'),token=(ghe_token, ''))
+        #resp=self.ghe_oauth.get('http://10.110.13.147:9090/auth/realms/brent/protocol/openid-connect/userinfo',token=(ghe_token,''))
```
Comment (Fokko) [inline airflow/contrib/auth/backends/keycloak_auth.py]: Please remove the comments

File: airflow/contrib/auth/backends/keycloak_auth.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+            if resp is None:
+                raise AuthenticationError(
+                    'Null response from GHE, denying access.'
+                )
+
+            ghe_token = resp['access_token']
+
+            username, email = self.get_ghe_user_profile_info(ghe_token)
+        except AuthenticationError:
+            _log.exception('')
```
Comment (Fokko) [inline airflow/contrib/auth/backends/keycloak_auth.py]: An empty log line

File: airflow/contrib/auth/backends/keycloak_auth.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+from flask import url_for, redirect, request
+
+from flask_oauthlib.client import OAuth
+
+from airflow import models, configuration, settings
+from airflow.configuration import AirflowConfigException
+
+import os
+os.environ['OAUTHLIB_INSECURE_TRANSPORT'] = '1'
```
Comment (Fokko) [inline airflow/contrib/auth/backends/keycloak_auth.py:34]: I think this should be disabled by default

File: airflow/contrib/auth/backends/keycloak_auth.py (hunk 1 above)
Comment (fisher-monkey) [reply airflow/contrib/auth/backends/keycloak_auth.py]: I have create a Jira ticket and simply describe the implementation, and this is the link: https://issues.apache.org/jira/browse/AIRFLOW-1836

the comments have been removed.

**Comment by ron819** [comment]
@fisher-monkey  @Fokko is there any progress with this?

**Comment by c4milo** [comment]
Would it be possible to make this more generic and support any OpenIDConnect/OAuth2 identity provider? being KeyCloak one of them.

**Comment by Fokko** [comment]
Very good suggestion @c4milo

**Comment by ron819** [comment]
@Fokko Does the suggestion to make it generic prevents from this to be merged or it's a feature request for future development? We do have other cases where non-generic PRs have been submitted. Example: https://github.com/apache/incubator-airflow/pull/3941

**Comment by Fokko** [comment]
I think it is for #3941 hard to implement it generic, i.e. have some logic to test the connection of an arbitrary connection. Implementing OAuth makes perfect sense. What do you suggest @ron819 ?

**Comment by weldpua2008** [comment]
@Fokko / @fisher-monkey Could you please inform if keycloak is supported in Airflow?
We started to use keycloak and I need to configure it.

**Comment by mik-laj** [comment]
@weldpua2008  You can configure any OAuth2 provider, but I am working on a Keycloak-proxy integration as it is much more secure. It's not public yet, but I have plans to share it with the community.

**Comment by farisabdulraheem** [comment]
How can i implement keycloak auth with airflow? @fisher-monkey  @Fokko  is there a simple way for that, any sample code available ?

**Comment by mik-laj** [comment]
this PR refers to the old and unsupported user interface. We support RBAC (Flask AppBuilser based) only now.

**Comment by farisabdulraheem** [comment]
@mik-laj  so how can i  use it with keycloak using rbac

**Comment by weldpua2008** [comment]
Any news?

**Pull request closed without merging**